### PR TITLE
feat: switch to elastic APM (part of #577)

### PR DESCRIPTION
**Pull request by MartinKolarik**
Migrates the current setup, including the existing custom metrics. New metrics will be added separately.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -117,14 +119,14 @@
...
 		"download:files": "tsx src/lib/download-files.ts",
 		"prepare": "npm run download:files; husky || echo 'Failed to install husky'",
-		"stats": "NEW_RELIC_ENABLED=false tsx probes-stats/known.ts",
-		"start": "node --max_old_space_size=3584 --max-semi-space-size=128 --experimental-loader newrelic/esm-loader.mjs -r newrelic dist/src/index.js",
-		"start:dev": "NODE_ENV=development FAKE_PROBE_IP=1 NEW_RELIC_ENABLED=false tsx src/index.ts",
-		"start:test": "NODE_ENV=test knex migrate:latest && NODE_ENV=test NEW_RELIC_ENABLED=false tsx src/index.ts",
+		"stats": "tsx probes-stats/known.ts",
+		"start": "ELASTIC_APM_CONFIG_FILE=elastic-apm-node.cjs node --max_old_space_size=3584 --max-semi-space-size=128 --experimental-loader elastic-apm-node/loader.mjs -r elastic-apm-node/start.js dist/src/index.js",
+		"start:dev": "NODE_ENV=development FAKE_PROBE_IP=1 ELASTIC_APM_CONFIG_FILE=elastic-apm-node.cjs tsx src/index.ts",
+		"start:test": "NODE_ENV=test knex migrate:latest && NODE_ENV=test ELASTIC_APM_CONFIG_FILE=elastic-apm-node.cjs tsx src/index.ts",
```
Comment (alexey-yarmosh) [inline package.json:129]: Is `ELASTIC_APM_CONFIG_FILE=elastic-apm-node.cjs` required here?

**Comment by alexey-yarmosh** [comment]
For development purposes it may be handy for logger to accept inputs with more arguments (`logger.info('message', { a: 1 }, { a: 2 }, { a: 3 });`), or different type (`logger.info({ a: 1 });`). But if it is so strict to make all messages similar I think that is fine.

**Review by alexey-yarmosh — APPROVED** [review]

File: elastic-apm-node.cjs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	version = fs.readFileSync(path.join(path.resolve(), 'data/LAST_API_COMMIT_HASH.txt'), 'utf8').trim();
+} catch {
+	version = require('./package.json').version;
+}
+
+module.exports = {
+	active: process.env.NODE_ENV === 'production',
+	serviceName: 'globalping-api',
+	serviceVersion: version,
+	logLevel: 'fatal',
```
[the remark this answers is not in the document]
Comment (MartinKolarik) [reply elastic-apm-node.cjs:16]: It's a log level of the agent's own internal messages which aren't particularly useful, that's why we log only the highest level.

File: package.json (hunk 1 above)
Comment (MartinKolarik) [reply package.json:129]: Yes because of the extension, it expects `js` otherwise.

**Comment by MartinKolarik** [comment]
It's strict on purpose. The first argument being the message makes it easier to know where the error is from without looking down the stack trace.

The limited number of arguments is to encourage doing e. g. `log(message, error, { req, res })` instead of `log(message, error, req, res)` so that it's clear what each value represents (because key names will be in the output).

It also enables customizing output for [specific keys](https://github.com/MartinKolarik/h-logger2/blob/13ba2beda491ad109c5dcae1a154277aef03ec59/src/Logger.js#L95) so when you pass in a request, for example, it only logs a few useful properties instead of hundreds of lines.

**Comment by alexey-yarmosh** [comment]
Ok, that makes sense 👍

**Comment by MartinKolarik** [comment]
I think we can move the datetime, which will make it much better. The level should be near the message and can have a different number of letters as well.

**Pull request merged**